### PR TITLE
`Programming exercises`: Improve error messages for connection issues in online code editor

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Students should not be confused when the connection issues to the Artemis server cause the online editor to malfunction. So the error messages should be improved.

### Description

Whenever an action is not possible in the code editor due to a connection issue an error is displayed at the top right telling the student about this problem.
Additionally when trying to load a file without an internet connection a blank file is shown instead of the last loaded file as before this PR. This blank file is read only.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise with online editor enabled

1. Log in to Artemis
2. Participate using the online editor
3. Loose internet connection by turning of wifi or stopping the server when testing locally
4. Interact with the code editor and check that all actions (submitting, loading an unloaded file, ...) cause an error message at the top right
5. Loading and editing existing files that were loaded before the connection loss should still be possible
6. Reconnect to the server
7. Check that any changes during the connection loss can now be successfully submitted

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Programming Exercise

1. Same as for normal exercises but within an exam

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [code-editor-ace.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5166/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/code-editor/ace/code-editor-ace.component.ts.html) | 80.83% | ✅ |
| [code-editor-container.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5166/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/code-editor/container/code-editor-container.component.ts.html) | 85.29% | ✅ |

### Screenshots

Bildschirmfoto 2023-07-13 um 11 56 36

### CODE REVIEW

**Review by rriyaldhi — DISMISSED** [review]
Code looks good.

File: src/main/webapp/app/exercises/programming/shared/code-editor/ace/code-editor-ace.component.ts [hunk 1]
```diff
@@ -107,7 +106,7 @@ export class CodeEditorAceComponent implements AfterViewInit, OnChanges, OnDestr
     isLoading = false;
     annotationsArray: Array<Annotation> = [];
     annotationChange: Subscription;
-    fileSession: { [fileName: string]: { code: string; cursor: { column: number; row: number } } } = {};
+    fileSession: { [fileName: string]: { code: string; cursor: { column: number; row: number }; loadingError: boolean } } = {};
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/shared/code-editor/ace/code-editor-ace.component.ts:109]: maybe refactor this into an explicit type? This line is quite long.

File: src/main/webapp/app/exercises/programming/shared/code-editor/container/code-editor-container.component.ts [hunk 2]
```diff
@@ -216,10 +217,18 @@ export class CodeEditorContainerComponent implements ComponentCanDeactivate {
...
      * Used by other components to display errors.
-     * The error must already be provided translated by the emitting component.
+     * @param error the translation key of the error that should be displayed
      */
     onError(error: any) {
-        this.alertService.error(`artemisApp.editor.errors.${error as string}`);
+        let errorTranslationKey: string;
+        const translationParams = { connectionIssue: '' };
+        if (typeof error !== 'string' || !error.includes(ConnectionError.message)) {
+            errorTranslationKey = error as string;
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/shared/code-editor/container/code-editor-container.component.ts:226]: Is this correct? You call `error as string` explicitly in cases where the error is not if type string.

Comment (JohannesStoehr) [reply src/main/webapp/app/exercises/programming/shared/code-editor/container/code-editor-container.component.ts]: Typecheck now removed. We also assumed before that the error is a string

**Review by YeetTheFirst21 — APPROVED** [review]
Tested on TS1, all seems to be working as expected

checked both normal coding exercises and exam mode.

**Review by MaximilianAnzinger — APPROVED** [review]
Code LGTM

**Review by RY997 — APPROVED** [review]
Tested on ts4, worked well.

**Review by Strohgelaender — APPROVED** [review]
code looks good to me

**Review by pal03377 — APPROVED** [review]
Offline mode and errors are working for me after some delay (see video).

I get errors for all operations except for deletion, where I get no feedback at all:
https://github.com/ls1intum/Artemis/assets/9006596/f36a3a46-c431-418a-903f-efcbfe18eb6c

I would be fine with this being solved in a follow-up though => approve ✅
Note: The deletion errors were visible for @aplr in Firefox, but not for me in Chrome.

**Review by tobias-lippert — APPROVED** [review]
Other than the issues @pal03377  mentioned, the improvements worked.

**Review by FelixTJDietrich — APPROVED** [review]
Tested on ts5 during the exam testing session. I got errors when not connected to wifi for all operations except the deletion of the file on both Safari and Chrome.

Note that rename and adding a file does explicitly state that the internet connection is bad.

**Review by basak-akan — APPROVED** [review]
Tested on TS5 exam mode testing session. Loading and editing existing files works as described.
However, I don't receive any error for deletion while offline as  @pal03377 mentioned.

**Pull request merged**